src/nrniv/partrans.cpp

```diff
@@ -1071,211 +1066,134 @@ extern size_t nrnbbcore_gap_write(const char* path, int* group_ids);
-      if (tml->index == type) {
-        gi[tid].gap_ml = tml->ml;
-      }
-    }
-  }
-
-  // gap info for targets, segregate into threads
-  for (int i=0; i < targets_->count(); ++i) {
-    int sid = sgid2targets_->item(i);
-    NrnThread* nt = (NrnThread*)target_pntlist_->item(i)->_vnt;
-    int tid = nt ? nt->id : 0;
-    Memb_list* ml = gi[tid].gap_ml;
-    int ix = targets_->item(i) - ml->data[0];
-
-    BBCoreGapInfo& g = gi[tid];
-    g.sid_target[g.ntar] = sid;
-    g.vpre_index[g.ntar] = ix;
-    g.ntar += 1;
-  }
-
-  // gap info for sources, segregate into threads.
-  for (int i=0; i < sgids_->count(); ++i) {
-    int sid = sgids_->item(i);
-    Node* nd = visources_->item(i);
-    assert(nd->extnode == NULL);
-    int tid = nd->_nt ? nd->_nt->id : 0;
-    int ix = nd->_v - nrn_threads[tid]._actual_v;
-    assert(ix >= 0 && ix < nrn_threads[tid].end);
-    BBCoreGapInfo& g = gi[tid];
-    g.sid_src[g.nsrc] = sid;
-    g.v_index[g.nsrc] = ix;
-    g.nsrc += 1;
-  }
-
-  // Sort each threads sources and targets so that the indices are
-  // monotonically increasing. (In fact the target indices will then be
-  // in memb_list order and that array will not have to be transferred.)
-  for (int tid=0; tid < nrn_nthread; ++tid) {
-    BBCoreGapInfo& g = gi[tid];
-    sidsort(g.sid_src, g.nsrc, g.v_index);
-    sidsort(g.sid_target, g.ntar, g.vpre_index);
+  if (gi) {
+    delete [] gi;
   }
-}
+  gi = new SetupTransferInfo[cn_nthread];
 
-static bool mycompar(const std::pair<int, int>& a, const std::pair<int, int>& b) {
-  return a.second < b.second;
-}
+  // info for targets, segregate into threads
+  if (targets_) {
+    for (int i=0; i < targets_->count(); ++i) {
+      sgid_t sid = sgid2targets_->item(i);
+      Point_process* pp = target_pntlist_->item(i);
+      NrnThread* nt = (NrnThread*)pp->_vnt;
+      int tid = nt ? nt->id : 0;
+      int type = pp->prop->type;
+      Memb_list& ml = *nt->_ml_list[type];
+      int ix = targets_->item(i) - ml.data[0];
 
-static void sidsort(int* sids, int cnt, int* indices) {
-  std::vector< std::pair< int, int> > si(cnt); // sids,indices
-  for (int i=0; i < cnt; ++i) {
-    si[i].first = sids[i];
-    si[i].second = indices[i];
+      SetupTransferInfo& g = gi[tid];
+      g.tar_sid.push_back(sid);
+      g.tar_type.push_back(type);
+      g.tar_index.push_back(ix);
+    }
   }
 
-  std::sort(si.begin(), si.end(), mycompar);
-
-  // copy back
-  for (int i=0; i < cnt; ++i) {
-    sids[i] = si[i].first;
-    indices[i] = si[i].second;
-  }
+  // info for sources, segregate into threads.
+  if (visources_) {
+    for (int i=0; i < sgids_->count(); ++i) {
+      sgid_t sid = sgids_->item(i);
+      Node* nd = visources_->item(i);
+      int tid = nd->_nt ? nd->_nt->id : 0;
+      int type = -1; // default voltage
+      int ix = 0; // fill below
+      NonVSrcUpdateInfo::iterator it = non_vsrc_update_info_.find(sid);
+      if (it != non_vsrc_update_info_.end()) {
```

Comment:
nitpick : could we add comment here about when `find` will fail / won't find an element?

Reply:
Added comments.